### PR TITLE
feat(rust): Add support for arrow binary type

### LINKED ISSUE #4903: Add support for Arrow Binary data type
Hey,

Currently binary data type is not supported but utf8 is supported.

It is much more efficient to store binary data directly in parquet files and doesn't make sense to convert the data to string just so it is queryable.

Adding support for Arrow Binary type would make this library more usable.

I can work on this if someone can give me a list of places I need to add Binary type to. I imagine it would be fairly easy to add this because utf8 is already supported

### PR DESCRIPTION
**Pull request by ozgrakkurt**
closes #4903

@ritchie46 can you check if this makes sense when you have time?

I redirected arrow dep to my fork temporarily

### CODE REVIEW

**Comment by ozgrakkurt** [comment]
@ritchie46 tested this using code I already had and it works for querying parquet files, filtering reading data etc. Just can't get the tests to compile. Also probably need to add some tests, can you point where I should add tests?

**Comment by ritchie46** [comment]
@ozgrakkurt thanks for the initial push on this. 

Can you feature gate this functionality? I only want this compiled if `dtype-binary` is activated.

Tests can be placed under `polars/tests/it/`.

**Comment by ozgrakkurt** [comment]
@ritchie46 could you do the feature gating if it is possible. Not sure how exactly to do it

**Comment by ozgrakkurt** [comment]
@ritchie46 can you check again?

**Comment by ozgrakkurt** [comment]
also closes #4877

File: polars/polars-core/src/chunked_array/ops/aggregate.rs [hunk 1]
```diff
@@ -804,6 +815,21 @@ impl QuantileAggSeries for Utf8Chunked {
     }
 }
 
+#[cfg(feature = "dtype-binary")]
+impl QuantileAggSeries for BinaryChunked {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/aggregate.rs:819]: These can be removed. The `Series` can just return `null` for the Binary dtype

File: polars/polars-core/src/chunked_array/ops/aggregate.rs [hunk 2]
```diff
@@ -692,6 +692,17 @@ impl VarAggSeries for Utf8Chunked {
     }
 }
 
+#[cfg(feature = "dtype-binary")]
+impl VarAggSeries for BinaryChunked {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/aggregate.rs:696]: These can be removed. The `Series` can just return `null` for the Binary dtype

File: polars/polars-core/src/chunked_array/ops/aggregate.rs [hunk 3]
```diff
@@ -849,6 +875,31 @@ impl ChunkAggSeries for Utf8Chunked {
     }
 }
 
+#[cfg(feature = "dtype-binary")]
+impl ChunkAggSeries for BinaryChunked {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/aggregate.rs:879]: These can be removed. The `Series` can just return `null` for the Binary dtype

File: polars/polars-core/src/chunked_array/ops/sort/mod.rs [hunk 4]
```diff
@@ -508,6 +508,137 @@ impl ChunkSort<Utf8Type> for Utf8Chunked {
     }
 }
 
+#[cfg(feature = "dtype-binary")]
+impl ChunkSort<BinaryType> for BinaryChunked {
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/ops/sort/mod.rs:512]: Should binary implement sort?

**Review by ritchie46 — COMMENTED** [review]
Thanks a lot for this @ozgrakkurt. I know this is a lot of work. I have left a few remarks on places where I think we can remove the code and let the `Series` simply return `null`. This often does this already automatically, so we can probably just remove the code.

I am only not sure we should implement `Ord` for binary? E.g. `min`, `max`, `sort`? Are byte slices `Ord` in rust? If so, we can follow.

File: polars/polars-core/src/chunked_array/ops/aggregate.rs (hunk 1 above)
Comment (ozgrakkurt) [reply polars/polars-core/src/chunked_array/ops/aggregate.rs:819]: I get this compilation error when I remove this:
```
error[E0277]: the trait bound `datatypes::BinaryType: datatypes::PolarsIntegerType` is not satisfied
   --> polars/polars-core/src/series/implementations/binary.rs:326:45
    |
326 |         QuantileAggSeries::median_as_series(&self.0)
    |         ----------------------------------- ^^^^^^^ the trait `datatypes::PolarsIntegerType` is not implemented for `datatypes::BinaryType`
    |         |
    |         required by a bound introduced by this call
    |
    = help: the following other types implement trait `datatypes::PolarsIntegerType`:
              datatypes::Int16Type
              datatypes::Int32Type
              datatypes::Int64Type
              datatypes::Int8Type
              datatypes::UInt16Type
              datatypes::UInt32Type
              datatypes::UInt64Type
              datatypes::UInt8Type
note: required for `chunked_array::ChunkedArray<datatypes::BinaryType>` to implement `chunked_array::ops::aggregate::QuantileAggSeries`
   --> polars/polars-core/src/chunked_array/ops/aggregate.rs:715:9
    |
715 | impl<T> QuantileAggSeries for ChunkedArray<T>
    |         ^^^^^^^^^^^^^^^^^     ^^^^^^^^^^^^^^^

error[E0277]: the trait bound `datatypes::BinaryType: datatypes::PolarsIntegerType` is not satisfied
   --> polars/polars-core/src/series/implementations/binary.rs:339:47
    |
339 |         QuantileAggSeries::quantile_as_series(&self.0, quantile, interpol)
    |         ------------------------------------- ^^^^^^^ the trait `datatypes::PolarsIntegerType` is not implemented for `datatypes::BinaryType`
    |         |
    |         required by a bound introduced by this call
    |
    = help: the following other types implement trait `datatypes::PolarsIntegerType`:
              datatypes::Int16Type
              datatypes::Int32Type
              datatypes::Int64Type
              datatypes::Int8Type
              datatypes::UInt16Type
              datatypes::UInt32Type
              datatypes::UInt64Type
              datatypes::UInt8Type
note: required for `chunked_array::ChunkedArray<datatypes::BinaryType>` to implement `chunked_array::ops::aggregate::QuantileAggSeries`
   --> polars/polars-core/src/chunked_array/ops/aggregate.rs:715:9
    |
715 | impl<T> QuantileAggSeries for ChunkedArray<T>
    |         ^^^^^^^^^^^^^^^^^     ^^^^^^^^^^^^^^^


```

**Comment by ozgrakkurt** [comment]
> Thanks a lot for this @ozgrakkurt. I know this is a lot of work. I have left a few remarks on places where I think we can remove the code and let the `Series` simply return `null`. This often does this already automatically, so we can probably just remove the code.
> 
> I am only not sure we should implement `Ord` for binary? E.g. `min`, `max`, `sort`? Are byte slices `Ord` in rust? If so, we can follow.

Yes if T is Ord any slice of T is ord as well I think. Those operations are useful in binary as well

**Comment by ritchie46** [comment]
Working on it!

**Comment by ritchie46** [comment]
Something went wrong. Reopend on #5122

**Pull request closed without merging**